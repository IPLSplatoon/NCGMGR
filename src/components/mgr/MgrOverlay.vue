<template>
    <dialog
        ref="dialog"
        @close="onClose"
        @cancel.prevent="onCancel"
        @click.self="onClick"
    >
        <div class="content">
            <slot />
        </div>
    </dialog>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref, watch } from 'vue'

async function animationsComplete(element: HTMLElement): Promise<unknown> {
    return Promise.allSettled(element.getAnimations().map(animation => animation.finished))
}

export default defineComponent({
    name: 'MgrOverlay',

    props: {
        visible: {
            type: Boolean,
            required: true
        },
        noBackgroundClose: {
            type: Boolean,
            default: false
        },
        maxWidth: {
            type: Boolean,
            default: false
        }
    },

    emits: ['update:visible'],

    setup (props, { emit }) {
        const dialog = ref<HTMLDialogElement | null>(null)

        async function onClose() {
            if (!dialog.value) return
            dialog.value.setAttribute('inert', '')
            await animationsComplete(dialog.value)
            dialog.value.close()
        }

        onMounted(() => {
            if (!props.visible) {
                dialog.value?.setAttribute('inert', '')
            } else {
                dialog.value?.showModal()
            }

            watch(() => props.visible, newValue => {
                if (newValue) {
                    dialog.value?.removeAttribute('inert')
                    dialog.value?.showModal()
                } else {
                    onClose()
                }
            })
        })

        return {
            dialog,
            onClose() {
                emit('update:visible', false)
            },
            onCancel() {
                if (!props.noBackgroundClose) {
                    onClose()
                }
            },
            onClick() {
                if (!props.noBackgroundClose) {
                    onClose()
                }
            }
        }
    }
})
</script>

<style lang="scss" scoped>
@use '../../styles/constants';

dialog {
    border: none;
    background-color: var(--ipl-bg-primary);
    color: var(--ipl-body-text-color);
    width: 85%;
    max-height: 80%;
    height: max-content;
    box-sizing: border-box;
    //margin: 0;
    padding: 0;
    inset: 0;
    display: block;
    position: fixed;
    // Unfortunately, if you want animation, you'll have to put up with this.
    z-index: 999999;
    transform: scale(0.9);
    opacity: 0;
    border-radius: constants.$border-radius-outer;

    &:not([open]) {
        pointer-events: none;
    }

    &[open] {
        animation: enter constants.$transition-duration-low ease-out forwards;

        &::backdrop {
            animation: backdrop-in constants.$transition-duration-low ease-out forwards;
        }

        &[inert] {
            animation: exit constants.$transition-duration-low ease-in forwards;

            &::backdrop {
                animation: backdrop-out constants.$transition-duration-low ease-out forwards;
            }
        }
    }

    &::backdrop {
        background: var(--ipl-page-overlay-color);
    }

    > .content {
        padding: 8px;
        box-sizing: border-box;
    }

    &:focus-visible {
        outline: var(--ipl-focus-outline-color) solid var(--ipl-focus-outline-width);
    }
}

@keyframes enter {
    to {
        transform: scale(1);
        opacity: 1;
    }
}

@keyframes exit {
    0% {
        transform: scale(1);
        opacity: 1;
    }

    100% {
        transform: scale(0.9);
        opacity: 0;
    }
}

@keyframes backdrop-in {
    0% {
        opacity: 0;
    }

    100% {
        opacity: 1;
    }
}

@keyframes backdrop-out {
    0% {
        opacity: 1;
    }

    100% {
        opacity: 0;
    }
}
</style>
