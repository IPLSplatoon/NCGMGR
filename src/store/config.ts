import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const configStore = createStore<Configuration>({
    state: {
        installPath: ''
    },
    mutations: {
        setInstallPath: (state, installPath: string) => {
            state.installPath = installPath
        },
        setState: (state, config: Configuration) => {
            Object.assign(state, config)
        }
    },
    actions: {
        load: store => {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            store.commit('setState', {
                ...store.state,
                ...config
            })
        },
        persist: store => {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(store.state))
        }
    }
})

export const configStoreKey: InjectionKey<Store<Configuration>> = Symbol()

export function useConfigStore (): Store<Configuration> {
    return useStore(configStoreKey)
}
