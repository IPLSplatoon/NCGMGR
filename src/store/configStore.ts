import { defineStore } from 'pinia'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const useConfigStore = defineStore('config', {
    state: () => ({
        installPath: ''
    } as Configuration),
    actions: {
        load () {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            this.$patch(config)
        },
        persist () {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(this.$state))
        }
    }
})
