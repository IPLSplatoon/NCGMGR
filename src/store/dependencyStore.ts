import { defineStore } from 'pinia'
import { isBlank } from '@iplsplatoon/vue-components'
import { getNodejsVersion } from '@/service/dependencyService'

interface DependencyStore {
    isLoading: boolean
    nodejs: {
        version: string | null
    }
}

export const useDependencyStore = defineStore('dependency', {
    state: () => ({
        isLoading: false,
        nodejs: {
            version: null
        }
    } as DependencyStore),
    getters: {
        hasNodejs: state => !isBlank(state.nodejs.version)
    },
    actions: {
        async checkVersions () {
            this.isLoading = true
            try {
                this.nodejs.version = await getNodejsVersion()
            } finally {
                this.isLoading = false
            }
        }
    }
})
