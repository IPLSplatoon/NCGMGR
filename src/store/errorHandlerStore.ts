import { App } from 'vue'
import { defineStore } from 'pinia'
import { generateId } from '@/util/generateId'

interface ErrorHandlerStore {
    recentErrors: Record<string, unknown>
}

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    } as ErrorHandlerStore),
    actions: {
        removeRecentError ({ key }: { key: string }): void {
            delete this.recentErrors[key]
        },
        handleError ({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err)

            const id = generateId()
            this.recentErrors[id] = err
        }
    }
})

export function setUpErrorHandler (app: App<unknown>): void {
    const store = useErrorHandlerStore()
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info })
    }
}
