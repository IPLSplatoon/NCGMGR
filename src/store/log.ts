import { LogEvent } from '@/types/log'
import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'
import { Event, listen, UnlistenFn } from '@tauri-apps/api/event'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    completed: Record<string, boolean>
}

const unlistenFns: Record<string, UnlistenFn> = {}

export const logStore = createStore<LogStore>({
    state: {
        lines: {},
        completed: {}
    },
    mutations: {
        insertLine (state, { line, key }: { line: LogEvent, key: string }) {
            if (!state.lines[key]) {
                state.lines[key] = []
            }
            state.lines[key].push(line)
        },
        reset (state, key: string) {
            state.lines[key] = []
            state.completed[key] = false
        },
        setCompleted (state, { key, completed }: { key: string, completed: boolean }) {
            state.completed[key] = completed
        }
    },
    actions: {
        listen (store, key: string) {
            if (unlistenFns[key]) {
                return
            }
            return listen(`log:${key}`, (event: Event<LogEvent>) => {
                store.commit('insertLine', { line: event.payload, key })
            }).then(unlisten => {
                unlistenFns[key] = unlisten
            })
        },
        unlisten (store, key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key]()
            }
        },
        logPromiseResult (store, { promise, key, noLogOnSuccess }: { promise: Promise<unknown>, key: string, noLogOnSuccess: boolean }) {
            promise.then(() => {
                if (!noLogOnSuccess) {
                    store.commit('insertLine', {
                        line: {
                            message: 'Success!',
                            type: 'success'
                        },
                        key
                    })
                }
            }).catch(e => {
                store.commit('insertLine', {
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
            }).finally(() => {
                store.commit('setCompleted', { key, completed: true })
            })
        }
    }
})

export const logStoreKey: InjectionKey<Store<LogStore>> = Symbol()

export function useLogStore (): Store<LogStore> {
    return useStore(logStoreKey)
}
