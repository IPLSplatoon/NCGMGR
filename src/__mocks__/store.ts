import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        }
    })
}
