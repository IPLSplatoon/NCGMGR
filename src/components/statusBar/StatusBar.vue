<template>
    <div>
        <ipl-space class="status-bar">
            <template
                v-for="(item, key) in items"
                :key="key"
            >
                <div
                    class="status-bar-item"
                    :class="{ highlighted: item.highlighted.value }"
                    @click="item.overlayVisible.value = true"
                    :id="`status-bar-item_${key}`"
                >
                    <font-awesome-icon :icon="item.icon" class="icon" />
                    <span class="text">{{ item.label }}</span>
                </div>
                <mgr-overlay v-model:visible="item.overlayVisible.value" :id="`status-item-dialog_${key}`">
                    <ipl-dialog-title
                        :title="item.label"
                        @close="item.overlayVisible.value = false"
                    />
                    <ipl-space class="m-t-8 dialog-component-wrapper">
                        <component :is="item.component" />
                    </ipl-space>
                </mgr-overlay>
            </template>
        </ipl-space>
    </div>
</template>

<script lang="ts">
import { defineComponent } from '@vue/runtime-core'
import { IplDialogTitle, IplSpace } from '@iplsplatoon/vue-components'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faWrench } from '@fortawesome/free-solid-svg-icons/faWrench'
import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons/faExclamationCircle'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import MgrOverlay from '@/components/mgr/MgrOverlay.vue'
import { computed, ref } from 'vue'
import DependencyChecker from '@/components/DependencyChecker.vue'
import { useDependencyStore } from '@/store/dependencyStore'
import ErrorList from '@/components/ErrorList.vue'
import { useErrorHandlerStore } from '@/store/errorHandlerStore'

library.add(faWrench, faExclamationCircle)

export default defineComponent({
    name: 'StatusBar',

    components: { DependencyChecker, IplDialogTitle, MgrOverlay, IplSpace, FontAwesomeIcon, ErrorList },

    setup () {
        const dependencyStore = useDependencyStore()
        const errorHandlerStore = useErrorHandlerStore()

        const items = {
            dependencyCheck: {
                icon: 'wrench',
                label: 'Dependency check',
                overlayVisible: ref(false),
                highlighted: computed(() => !dependencyStore.isLoading && !dependencyStore.hasNodejs),
                component: 'DependencyChecker'
            },
            errorLog: {
                icon: 'exclamation-circle',
                label: 'Error log',
                overlayVisible: ref(false),
                highlighted: computed(() => Object.keys(errorHandlerStore.recentErrors).length > 0),
                component: 'ErrorList'
            }
        }

        return {
            items
        }
    }
})
</script>

<style lang="scss" scoped>
@import 'src/styles/colors';

@keyframes highlight {
    0%  { color: var(--text-color) }
    49% { color: var(--text-color) }
    50% { color: $red }
    100% { color: $red }
}

.status-bar {
    display: flex;

    > .status-bar-item {
        padding: 2px 4px;
        cursor: pointer;
        color: var(--text-color);
        display: flex;
        align-items: center;

        > .icon {
            font-size: 1.2em;
        }

        > .text {
            display: none;
            margin-left: 4px;
        }

        &:hover > .text {
            display: unset;
        }

        &.highlighted > .icon {
            animation: highlight 2s infinite;
        }
    }
}

#status-item-dialog_errorLog .dialog-component-wrapper {
    padding: 0;
    overflow: hidden;
}
</style>
