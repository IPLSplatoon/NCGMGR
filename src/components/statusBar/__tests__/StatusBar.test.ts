import { createTestingPinia, TestingPinia } from '@pinia/testing'
import { config, mount } from '@vue/test-utils'
import StatusBar from '@/components/statusBar/StatusBar.vue'
import { useDependencyStore } from '@/store/dependencyStore'
import MgrOverlay from '@/components/mgr/MgrOverlay.vue'
import { IplDialogTitle } from '@iplsplatoon/vue-components'

describe('StatusBar', () => {
    let pinia: TestingPinia

    beforeEach(() => {
        pinia = createTestingPinia()
        config.global.plugins = [pinia]
    })

    config.global.stubs = {
        IplDialogTitle: true,
        FontAwesomeIcon: true,
        DependencyChecker: true
    }

    it('matches snapshot', () => {
        const wrapper = mount(StatusBar)

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('highlights dependency check icon when dependencies are missing', () => {
        const dependencyStore = useDependencyStore()
        dependencyStore.isLoading = false
        dependencyStore.nodejs.version = null
        const wrapper = mount(StatusBar)

        expect(wrapper.get('#status-bar-item_dependencyCheck').classes()).toContain('highlighted')
    })

    it.each([
        [true, null],
        [true, '1.2.3'],
        [false, '1.2.4']
    ])('does not highlight dependency check when dependency loading state is %s and nodejs version is %s', (isLoading, nodeVersion) => {
        const dependencyStore = useDependencyStore()
        dependencyStore.isLoading = isLoading
        dependencyStore.nodejs.version = nodeVersion
        const wrapper = mount(StatusBar)

        expect(wrapper.get('#status-bar-item_dependencyCheck').classes()).not.toContain('highlighted')
    })

    it('opens dialog when clicking on status bar item', async () => {
        const wrapper = mount(StatusBar)

        await wrapper.get('#status-bar-item_dependencyCheck').trigger('click')

        expect(wrapper.getComponent<typeof MgrOverlay>('#status-item-dialog_dependencyCheck').vm.visible).toEqual(true)
    })

    it('handles closing opened dialogs', async () => {
        const wrapper = mount(StatusBar)

        await wrapper.get('#status-bar-item_dependencyCheck').trigger('click')
        wrapper
            .getComponent<typeof MgrOverlay>('#status-item-dialog_dependencyCheck')
            .getComponent<typeof IplDialogTitle>('ipl-dialog-title-stub').vm.$emit('close')
        await wrapper.vm.$nextTick()

        expect(wrapper.findComponent<typeof MgrOverlay>('#status-item-dialog_dependencyCheck').exists()).toEqual(false)
    })
})
