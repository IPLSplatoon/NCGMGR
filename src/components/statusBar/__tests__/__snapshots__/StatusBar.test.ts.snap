// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBar matches snapshot when error log is disabled 1`] = `
"<div data-v-d255ffee="" class="ipl-space color-secondary status-bar">
  <div id="status-bar-item_dependencyCheck" class="status-bar-item highlighted">
    <font-awesome-icon-stub icon="wrench" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon"></font-awesome-icon-stub><span class="text">Settings</span>
  </div>
  <!--v-if-->
</div>"
`;

exports[`StatusBar matches snapshot when error log is enabled 1`] = `
"<div data-v-d255ffee="" class="ipl-space color-secondary status-bar">
  <div id="status-bar-item_dependencyCheck" class="status-bar-item highlighted">
    <font-awesome-icon-stub icon="wrench" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon"></font-awesome-icon-stub><span class="text">Settings</span>
  </div>
  <!--v-if-->
  <div id="status-bar-item_errorLog" class="status-bar-item">
    <font-awesome-icon-stub icon="exclamation-circle" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon"></font-awesome-icon-stub><span class="text">Error log</span>
  </div>
  <!--v-if-->
</div>"
`;
